src/main/java/io/confluent/kafkarest/SimpleConsumerManager.java

```diff
@@ -0,0 +1,233 @@
+            messageAndOffset.message(), messageAndOffset.offset(),
+            binaryDecoder, binaryDecoder);
+    return new BinaryConsumerRecord(messageAndMetadata.key(), messageAndMetadata.message(),
+        partitionId, messageAndOffset.offset());
+  }
+
+  private AvroConsumerRecord createAvroConsumerRecord(final MessageAndOffset messageAndOffset,
+                                                      final String topicName,
+                                                      final int partitionId) {
+    final MessageAndMetadata<Object, Object> messageAndMetadata =
+        new MessageAndMetadata<Object, Object>(topicName, partitionId,
+            messageAndOffset.message(), messageAndOffset.offset(),
+            avroDecoder, avroDecoder);
+    return new AvroConsumerRecord(
+        AvroConverter.toJson(messageAndMetadata.key()).json,
+        AvroConverter.toJson(messageAndMetadata.message()).json,
+        partitionId, messageAndOffset.offset());
+  }
+
+  private ConsumerRecord createConsumerRecord(final MessageAndOffset messageAndOffset,
+                                              final String topicName,
+                                              final int partitionId,
+                                              final EmbeddedFormat embeddedFormat) {
+
+    switch (embeddedFormat) {
+      case BINARY:
+        return createBinaryConsumerRecord(messageAndOffset, topicName, partitionId);
+
+      case AVRO:
+        return createAvroConsumerRecord(messageAndOffset, topicName, partitionId);
+
+      default:
+        throw new RestServerErrorException("Invalid embedded format for new consumer.",
+            Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
+    }
+  }
+
+  private ByteBufferMessageSet fetchRecords(final String topicName,
+                                            final int partitionId,
+                                            final long offset,
+                                            final SimpleFetcher simpleFetcher) {
+
+    final ConsumerConfig consumerConfig = simpleConsumerFactory.getConsumerConfig();
+
+    final Map<TopicAndPartition, PartitionFetchInfo> requestInfo =
+        new HashMap<TopicAndPartition, PartitionFetchInfo>();
+
+    requestInfo.put(new TopicAndPartition(topicName, partitionId),
+        new PartitionFetchInfo(offset, consumerConfig.fetchMessageMaxBytes()));
+
+    final int corId = correlationId.incrementAndGet();
+
+    final FetchRequest req =
+        new FetchRequest(
+            corId,
+            simpleFetcher.clientId(),
+            consumerConfig.fetchWaitMaxMs(),
+            consumerConfig.fetchMinBytes(),
+            requestInfo);
+
+    final FetchResponse fetchResponse = simpleFetcher.fetch(req);
+
+    if (fetchResponse.hasError()) {
+      final short kafkaErrorCode = fetchResponse.errorCode(topicName, partitionId);
+      throw Errors.kafkaErrorException(new Exception("Fetch response contains an error code: " + kafkaErrorCode));
```

Comment:
Do you know off hand how many different types of errors FetchResponses can actually include? If we can get 1-3 that cover 80+% of the cases seen in practice, it can be better to give them more specific error codes (added to Errors.java) since they can then check the error programmatically to make decisions like whether they should retry or just give up. My quick check suggests a few that could be returned: UnknownTopicOrPartitionException, NotLeaderForPartitionException, ReplicaNotAvailableException, OffsetOutOfRangeException. Although those might not all be common. 

We also might want to respond differently to some of them than just passing them to the user. For example, OffsetOutOfRangeException is retriable and we might want to do a few rounds of retries. NotLeaderForPartition is another good example -- that's an InvalidMetadataException in the new client exception hierarchy, which would probably mean we should retry again completely from scratch, including the metadata lookup (which might mean an invalidation if we introduce a cache).